File: Codes/Yangkun Huang/String/38. Count and Say.cpp
string countAndSay(int n) {
	string s = "1";
	for (int i = 1; i<n; i++)
	{
		string temp;
		for (int j = 0, count = 1; j<s.length(); j++)
			if (s[j] == s[j + 1])
				count++;
			else
			{
				temp.insert(temp.end(), '0' + count);
				temp.insert(temp.end(), s[j]);
				count = 1;
			}
		s = temp;
	}
	return s;
}

string recursion(int n)
{
	if (n == 1)
		return "1";
	else
	{
		string s, temp;
		s = countAndSay(n - 1);
		for (int j = 0, count = 1; j<s.length(); j++)
			if (s[j] == s[j + 1])
				count++;
			else
			{
				temp.insert(temp.end(), '0' + count);
				temp.insert(temp.end(), s[j]);
				count = 1;
			}
		return temp;
	}
}

File: Codes/Bingran Li/String/38. Count and Say.java
public class Solution {
    public String countAndSay(int n) {
        
        // what the hell.....not understanding at first
        // solution 1: -----> 80%
        if(n==1)
            return "1";
            
        String prev = countAndSay(n-1);
        // say the prev string
        char [] prevArray = prev.toCharArray();
        
        int outputNum =prevArray[0]-'0';
        int count =0;
        StringBuilder myString = new StringBuilder(); // good choice
        
        for(int i =0;i<prevArray.length;i++,count++){
            if(prevArray[i]-'0'!=outputNum){
                
                myString.append(count);
                myString.append(outputNum);
                outputNum = prevArray[i]-'0';
                count = 0; // used to be 1----> changes are likely to bring mistakes
            }
        }
         myString.append(count);
         myString.append(outputNum);
        
        return myString.toString();
        
        // solution 2: iterative -----refer HUANG
    }
}
